react/features/mobile/external-api/middleware.js

```diff
@@ -30,14 +31,19 @@ MiddlewareRegistry.register(store => next => action => {
 
         // XXX Certain CONFERENCE_FAILED errors are recoverable i.e. they have
         // prevented the user from joining a specific conference but the app may
-        // be able to eventually join the conference. For example, the app will
-        // ask the user for a password upon
-        // JitsiConferenceErrors.PASSWORD_REQUIRED and will retry joining the
-        // conference afterwards. Such errors are to not reach the native
+        // be able to eventually join the conference.
+        //
+        // - JitsiConferenceErrors.AUTHENTICATION_REQUIRED: the app will ask the
+        //   user for credentials before joining the conference and will retry
+        //   joining once they are provided.
+        // - JitsiConferenceErrors.PASSWORD_REQUIRED: the app will ask the user
+        //   for a password and will retry joining the conference afterwards.
+        //
+        // Such errors are to not reach the native
         // counterpart of the External API (or at least not in the
         // fatality/finality semantics attributed to
         // conferenceFailed:/onConferenceFailed).
-        if (!error.recoverable) {
+        if (!isConferenceErrorRecoverable(error)) {
```

Comment:
This might be a matter of personal preference whether it's better to use a function or have a field set on the action. Probably there could be a problem with the fields solution, because some middleware could read the action before the field was set. If we would go with fields then in my opinion the check and field assignment should happen when the action is being emitted (it could call the new method that was introduced).

I'm leaving this comment only, because I don't know if @lyubomir decides if it's ok to go with function or action field.

Reply:
Personally I would prefer to have a 'recoverable' field set on the action instead of having to call the function in all places.

Reply:
You mentioned it, but to reiterate: The error object doesn’t have a recoverable field. It’s currently injected by the password-lock middleware and I find that approach incorrect on 2 accounts:

- other middlewares won’t see the attribute unless the run next(action) in order for this middleware to inject it.
- IMHO it’s not up to a feature to decide if an error is recoverable or not, that’s a decision we take globally.

Now imagine 5 different middlewares setting a flag on the error for 5 different reasons. With the approach on this PR the logic is in a si file place and it’s IMHO easier to track and maintain in the future.

Reply:
I see that @lyubomir has merged this PR in some other form. I talked to him and I agree with the choices he made.

> other middlewares won’t see the attribute unless the run next(action) in order for this middleware to inject it.

As @lyubomir explained error should not be process until it propagates across the app and is reduced, so you should not be doing that until next(action) is called.

> it’s not up to a feature to decide if an error is recoverable or not, that’s a decision we take globally.

With the version that @lyubomir merged it is the feature that recovers from error that decides if the error is recoverable (the feature knows it can handle that, so it marks it as recoverable).

Reply:
Thank you, @paweldomas and @saghul! I'm open to discussion. Today I need the most conservative, smallest, and safest modifications in order to unblock the testing and releasing of the SDK and the app.